tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -94,14 +114,257 @@ public int Serialize(ref byte[] bytes, int offset, Span value, IFormatterResolve
+            }
+
+            // add "_dd.origin" tag to all spans
+            string origin = span.Context.Origin;
+            if (!string.IsNullOrEmpty(origin))
+            {
+                count++;
+                offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _originNameBytes);
+                offset += MessagePackBinary.WriteString(ref bytes, offset, origin);
+            }
+
+            if (count > 0)
+            {
+                // Back-patch the count. End of "meta" dictionary. Do not add any string tags after this line.
+                MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, countOffset, (uint)count);
+            }
+
+            return offset - originalOffset;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteTag(ref byte[] bytes, ref int offset, string key, string value, ITagProcessor[] tagProcessors)
+        {
+            if (tagProcessors is not null)
+            {
+                for (var i = 0; i < tagProcessors.Length; i++)
+                {
+                    tagProcessors[i]?.ProcessMeta(ref key, ref value);
+                }
+            }
+
+            offset += MessagePackBinary.WriteString(ref bytes, offset, key);
+            offset += MessagePackBinary.WriteString(ref bytes, offset, value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void WriteTag(ref byte[] bytes, ref int offset, byte[] keyBytes, string value, ITagProcessor[] tagProcessors)
+        {
+            if (tagProcessors is not null)
+            {
+                string key = null;
+                for (var i = 0; i < tagProcessors.Length; i++)
+                {
+                    tagProcessors[i]?.ProcessMeta(ref key, ref value);
+                }
+            }
+
+            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, keyBytes);
+            offset += MessagePackBinary.WriteString(ref bytes, offset, value);
+        }
+
+        // METRICS
+
+        private int WriteMetrics(ref byte[] bytes, int offset, Span span, ITags tags, ITagProcessor[] tagProcessors)
+        {
+            int originalOffset = offset;
+
+            // Start of "metrics" dictionary. Do not add any numeric tags before this line.
+            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, _metricsBytes);
+
+            int count = 0;
+
+            // We don't know the final count yet, write a fixed-size header and note the offset
+            var countOffset = offset;
+            offset += MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, offset, 0);
+
+            // Write span metrics
+            var tagWriter = new TagWriter(this, tagProcessors, bytes, offset);
+            tags.EnumerateMetrics(ref tagWriter);
+            bytes = tagWriter.Bytes;
+            offset = tagWriter.Offset;
+            count += tagWriter.Count;
+
+            if (span.IsRootSpan)
```

Comment:
Same comment above about "root span" vs "first span in every chunk".

Reply:
I prefer to do all behaviour changes in a different PR. The scope for this PR was only separation of concerns.